### PR TITLE
Subword recipe for CSJ

### PR DESCRIPTION
**Pull request by jnishi**
This is a CSJ recipe using subword tokens. You can run this script by $ ./local/run_spm.sh.

This PR is refinement of https://github.com/espnet/espnet/pull/306.

### CODE REVIEW

**Comment by sw005320** [comment]
If you confirm some performance improvement, I want to merge it. Can you tune the number of units a little bit more?

**Comment by chenzhehuai** [comment]
@jnishi hi, I recently run ./local/run_spm.sh. in librispeech with 2K-10K units. The model seems to fail in converging. I am curious whether you have some bpe-based results in this corpus?

**Comment by sw005320** [comment]
@chenzhehuai, my suggestion is to reduce the number of encoder layers (say 6 or 5). We also observed such issues. Also, I had discussions with people in Google and RWTH and found that we should use shallow encoders with larger units (e.g., 3 layers but 1024 units). I'm now checking such configurations, but if you also play with them, and report your results, that is great.

**Comment by chenzhehuai** [comment]
@sw005320 I will. What do you think about their pre-training techs?

**Comment by sw005320** [comment]
It makes a training recipe bit complicated, and it is not first priority.

**Comment by chenzhehuai** [comment]
I tune number and width of encoder layers, nbpe units, sub-sampling and mtlalpha.
My current best setup is as below. The performance is 5% better than 1char system with the same setups and 10% better than 1char system with the setup of run.sh: 
train_100_blstmp_e6_subsample1_2_2_1_1_unit1024_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs50_mli800_mlo150_unigram200
Would you like me to give a PR? @sw005320

**Comment by sw005320** [comment]
Sounds great!
I also recently tuned the librispeech and made a PR in #354. In my setup, I have changed all units to 1024, but it may not be required, and your modification would be enough. Can you check #354, and see how we can do more?

**Comment by jnishi** [comment]
@sw005320 

I run the script with large size of encoder layer, but cannot improve performance.

|bpe size | # of encoder layers | size of encoder layers | network | backend | # GPUs | batchsize | e1 | e2 | e3 |
|-- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
|no BPE | 6 | 320 | vggblstmp | chainer | 1 | 30 | 8.7 | 6.2 | 6.9 |
|5000 | 4 | 1024 | vggblstm | chainer | 1 | 32 | 8.5 | 6.3 | 7.2 |
|8000 | 4 | 1024 | vggblstm | chainer | 1 | 32 | 9 | 6.9 | 7 |

I will try vggblstmp. Do you have any ideas of size of projection layer if size of encoder layer is 1024?

**Comment by sw005320** [comment]
Did you also change the number of units in a decoder?

**Comment by jnishi** [comment]
No, I didn't. The parameter of the decoder is the same as before.

**Comment by sw005320** [comment]
Cool!
Can you update `RESULTS` with your numbers? 
Then I'll merge it.

Can you also run the same architecture with the char system?
I'm expecting that the performance improvement comes from the architecture, not BPE (because the Japanese character is already BPEed (?)).

**Comment by sw005320** [comment]
Sorry to ask things separately.
Can you also update your results at `README.md`?

**Comment by jnishi** [comment]
@sw005320 
I updated `RESULTS` default parameters of `run_spm.sh`, and `README.md`. Can you check it?

I'm going to run char system with the same parameters.

File: egs/csj/asr1/local/run_spm.sh [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+# You can skip this and remove --rnnlm option in the recognition (stage 5)
+lmexpdir=exp/train_rnnlm_2layer_bs256_${bpemode}${nbpe}
+mkdir -p ${lmexpdir}
+if [ ${stage} -le 3 ]; then
+    echo "stage 3: LM Preparation"
+    lmdatadir=data/local/lm_train_${bpemode}${nbpe}
+    mkdir -p ${lmdatadir}
+    spm_encode --model=${bpemodel}.model --output_format=piece < data/lang_char/input.txt | perl -pe 's/\n/ <eos> /g' \
+        > ${lmdatadir}/train.txt
+    cut -f 2- -d" " data/${train_set}/text | spm_encode --model=${bpemodel}.model --output_format=piece | perl -pe 's/\n/ <eos> /g' \
```
Comment (sw005320) [inline egs/csj/asr1/local/run_spm.sh]: This should be `${train_set}` -> `${train_dev}`.
Librispeech recipe may also be like that.
Please fix it.

**Comment by kan-bayashi** [comment]
I tried shallow and wide setting with pytorch backend, which leads to the improvements.

```bash
$ rg -e Avg exp/train_nodup_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d1_unit1024_location_adim1024_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_eval*0.3*0.3/result.txt | sort | sed -E 's/ +/ /g'
exp/train_nodup_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d1_unit1024_location_adim1024_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3/result.txt:| Sum/Avg | 1272 43897 | 93.3 4.6 2.1 1.0 7.7 60.5 |
exp/train_nodup_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d1_unit1024_location_adim1024_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_eval2_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3/result.txt:| Sum/Avg | 1292 43623 | 95.1 3.7 1.3 0.7 5.7 58.9 |
exp/train_nodup_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d1_unit1024_location_adim1024_aconvc10_aconvf100_mtlalpha0.5_adadelta_bs30_mli800_mlo150/decode_eval3_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3/result.txt:| Sum/Avg | 1385 28225 | 94.9 3.7 1.4 1.1 6.2 42.4 |
```

**Comment by jnishi** [comment]
fixed path to validation dataset of language model.

**Comment by jnishi** [comment]
In my experiments, character model is improved in shallow and wide model.
Performance with fixing path to validation data of LM is listed below.

|bpe size | # of encoder layers | size of encoder layers |size of a decoder layer| network | backend | # GPUs | batchsize | e1 | e2 | e3 |
|-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
| no BPE | 4 | 1024 | 1024 | vggblstm | chainer | 1 | 32 | 7.3 | 5.3 | 5.9 |
| 5000 | 4 | 1024 |1024 | vggblstm | chainer | 1 | 32 | 8.0 | 6.1 | 6.1 |

**Comment by kan-bayashi** [comment]
@jnishi Very good results. Why don't you update default parameters in `run.sh` and add to `RESULTS`?

**Comment by jnishi** [comment]
I have updated `RESULTS` and default parameters of `run.sh`.

**Comment by sw005320** [comment]
Please also update `README.md`. Then, I'll merge it. It is unfortunate that BPE does not work for CSJ, but we can still keep to maintain it as `local/run_spm.sh`

**Comment by jnishi** [comment]
updated `README.md`.

**Pull request merged**